Add the kind primary

ekinanp opened the pull request:
Resolves https://github.com/puppetlabs/wash/issues/404

ekinanp committed:
Unmarshal entry schemas in their stree representation

The reasoning behind this is outlined in the comments of
apitypes.EntrySchema#UnmarshalJSON. The TLDR is that unmarshalling entry
schemas this way makes the way we think about entry schemas symmetric
with what the user sees when running the stree command. For example, the
stree representation lets us distinguish "Docker volume files" from "Docker container files".
The previous unmarshalling would (incorrectly) consider them the same entry
because they both are implemented using the same helper class (volume.dir).

Signed-off-by: Enis Inan <enis.inan@puppet.com>

ekinanp committed:
Refactor the entry predicate binary operators to implement Negate

We'll need this when implementing the kind primary.

Signed-off-by: Enis Inan <enis.inan@puppet.com>

ekinanp committed:
Add kind primary to filter on specific kinds of entries

Resolves https://github.com/puppetlabs/wash/issues/404

Signed-off-by: Enis Inan <enis.inan@puppet.com>

ekinanp commented on api/types/entrySchema.go:
```diff
@@ -47,49 +100,80 @@ func (s *EntrySchema) UnmarshalJSON(data []byte) error {
-// TypeID returns the entry's type ID. This should be unique.
+// Path returns the unique path to this specific entry's schema
+// (relative to the stree root). The path consists of
+//    <root_label>/<parent1_label>/.../<label>
```
This format makes the kind primary a bit annoying to use from the mountpoint. For example, if I want to find only Docker containers, I'll need to do something like `find -k "mountpoint/docker*container"` (I can't do `find -k "*docker*container"` because that will return containers in the `mock_docker` plugin).

I've been thinking about some alternatives:
1. Change `root_label` => `.` (which would indicate that the path is relative to the root). Then the previous example becomes something like `find -k "./docker*container"`. Problem with this is if one has multiple paths, e.g. `find <path1> <path2>`, it makes it impossible to filter on the kind of `<path1>`, `<path2>`.

2. Add some munging to the glob where if it starts with `.`, we will consider that the kind of `<path1>`, `<path2>`, depending on which paths' being walked. So, `./docker*container` would be munged to `mountpoint/docker*container`.

3. Change the mountpoint label to `.`, but keep everything else as-is. Think this makes the most sense, since the mountpoint label isn't being used anywhere.

MikaelSmith replied:
What about treating an absolute path as the start of the schema path? So `/docker*container`? Not sure the `.` is necessary to clarify that description.

ekinanp replied:
Yeah, could also do that too. Would that resolve the issue in (1)?

MikaelSmith replied:
It's not clear to me how any of them enable that. Maybe I'm getting confused by what `mountpoint` is. I assumed that would normally be whatever we mount Wash in, e.g. `mnt28139`.

Maybe the `/` is unnecessary because you're expecting a full match. For the initial example, why doesn't `docker/*container` match all (and only) containers? For example (1), could you give some concrete examples?

ekinanp replied:
`mountpoint` is plugin.Registry's label.

In the initial example, the schema paths of entries in the `docker` and the `mock_docker` plugin will start with `mountpoint/docker` and `mountpoint/mock_docker`, respectively (assuming we're invoking `find` in the mountpoint -- the mountpoint's entry is plugin.Registry). Since users should not care about the mountpoint label, they'd need to do something like `*docker*container` to match Docker containers -- `*docker` matches both `docker` and `mock_docker`.

The `<path1>`, `<path2>` stuff is a bit theoretical. Here's one example,
```
find <ec2_instance> <s3_bucket> -k 'instance' -o -k 'bucket'
```

Here you're filtering on entries of kind EC2 Instance or kind Bucket, so you'd print out `<ec2_instance>` and `<s3_bucket>`. The schema paths of `<ec2_instance>` and `<s3_bucket>` is (currently) `instance` and `bucket`, which is why this works. Munging the schema paths to `./` or `/` would lose their kinds, so the above expression would be incorrect.

Even though the above example is theoretical, I'd still like to get it right because there's no reason why you shouldn't be able to do that (all other predicates make it possible for you to print `<ec2_instance>` and `<s3_bucket>`, for example). 

I'm starting to like Option 3 more.

ekinanp replied:
Note that you can also do
```
find docker -k '*container'
```

to filter out Docker containers. We could also document that in the kind primary's documentation.

MikaelSmith replied:
Ok, so the problem is you're using glob somewhat unconventionally. In Linux, the `*` doesn't normally match separators and you have to provide a full path. Somewhere the `**` pattern was added to match directory trees. You're not passing any separators to the glob library, so it's matching them with a simple `*`.

My inclination would be to include separators because I think it avoids some common ambiguities. So your examples would be (omitting out the mountpoint label because we shouldn't use it)
- `docker/**/container`
- `**/instance`

The way you're configuring the glob calls, those examples have to be
- `/docker/*/container` to avoid ambiguities such as a plugin named `mock_docker` or `docker_new`, and new things like `dead_container` (if you're trying to be that specific). In practice maybe `/docker/*container` is fine.
- I have no idea how you're making `instance` work, because the glob library still needs a full match. I don't think the code you have here supports it, so you'd need to do something like `*instance` or `*/instance`.

The details about what the mountpoint should be called seem unimportant because glob works on full paths, and we should just exclude the mountpoint completely.

MikaelSmith replied:
Ok, I'm caught up. `find -path` is a pattern that treats separators as normal characters, so it doesn't behave the same as shell globbing. That's the behavior we're basing this new primary on.

MikaelSmith commented:
Hmm, I wonder why Travis hasn't started yet.

ekinanp commented on cmd/internal/find/primary/kind.go:
```diff
@@ -0,0 +1,70 @@
+
+// TODO: Fill this description out
+const kindDetailedDescription = `
+(-k|-kind) glob
```
I'll flesh this out once https://github.com/puppetlabs/wash/pull/449#discussion_r313008477 is addressed.

ekinanp commented on api/types/entrySchema.go:
```diff
@@ -38,7 +39,59 @@ func (s EntrySchema) MarshalJSON() ([]byte, error) {
 }
 
-// UnmarshalJSON unmarshals the entry's schema from JSON.
+// UnmarshalJSON unmarshals the entry's schema from JSON. EntrySchemas
```
Hmm, I think this could move to the EntrySchema type declaration b/c [fs/schema's response is a different type](https://github.com/puppetlabs/wash/blob/master/api/schema.go#L16).

ekinanp replied:
Regardless, I consider the refactor here a bug-fix (even though the previous unmarshalling didn't cause any issues in Wash's normal operation).

MikaelSmith commented on cmd/internal/find/types/testdata/triangle.json:
```diff
@@ -1,15 +1,18 @@
 {
     "A": {
+        "label": "A",
```
Is there an example where `label` isn't redundant?

ekinanp replied:
Yes, see https://github.com/ekinanp/wash/blob/kind-primary-round2/cmd/internal/find/types/testdata/real_world_aws.json

ekinanp committed:
Change glob => pattern

The latter name avoids conflating glob patterns with shell globs, which
work a bit differently (e.g. they do not match the "/" separator).

Signed-off-by: Enis Inan <enis.inan@puppet.com>

ekinanp commented:
Ok, this is ready for review.
